src/Raven.Server/Documents/Sharding/Operations/BulkInsert/ShardedBulkInsertOperation.cs

```diff
@@ -0,0 +1,226 @@
+    {
+        OperationId = id;
+        _skipOverwriteIfUnchanged = skipOverwriteIfUnchanged;
+        _databaseContext = databaseContext;
+        _context = context;
+        _token = token;
+
+        _streamExposerPerShard = new BulkInsertOperation.BulkInsertStreamExposerContent[databaseContext.ShardCount];
+        _currentWriters = new MemoryStream[databaseContext.ShardCount];
+        _backgroundWriters = new MemoryStream[databaseContext.ShardCount];
+        _asyncWrites = new Task[databaseContext.ShardCount];
+        _first = new bool[databaseContext.ShardCount];
+
+        for (int i = 0; i < databaseContext.ShardCount; i++)
+        {
+            _streamExposerPerShard[i] = new BulkInsertOperation.BulkInsertStreamExposerContent();
+            _currentWriters[i] = new MemoryStream();
+            _backgroundWriters[i] = new MemoryStream();
+            _asyncWrites[i] = Task.CompletedTask;
+            _first[i] = true;
+        }
+
+        _disposeOnce = new DisposeOnceAsync<SingleAttempt>[databaseContext.ShardCount];
+
+        for (int i = 0; i < databaseContext.ShardCount; i++)
+        {
+            var shardNumber = i;
+
+            _disposeOnce[shardNumber] = new DisposeOnceAsync<SingleAttempt>(async () =>
+            {
+                try
+                {
+                    if (_streamExposerPerShard[shardNumber].IsDone)
+                        return;
+
+                    if (_requestBodyStreamPerShard?[shardNumber] != null)
+                    {
+                        _currentWriters[shardNumber].WriteByte((byte)']');
+                        _currentWriters[shardNumber].Flush();
+                        await _asyncWrites[shardNumber];
+
+                        _currentWriters[shardNumber].TryGetBuffer(out var buffer);
+                        await _requestBodyStreamPerShard[shardNumber].WriteAsync(buffer.Array, buffer.Offset, buffer.Count, _token);
+                        await _requestBodyStreamPerShard[shardNumber].FlushAsync(_token);
+                    }
+
+                    _streamExposerPerShard[shardNumber].Done();
+                }
+                finally
+                {
+                    _streamExposerPerShard[shardNumber]?.Dispose();
+                }
+            });
+        }
+    }
+
+    public CompressionLevel CompressionLevel { get; set; } = CompressionLevel.NoCompression;
+
+    public HttpResponseMessage Combine(Memory<HttpResponseMessage> results) => null;
+
+    public RavenCommand<HttpResponseMessage> CreateCommandForShard(int shardNumber)
+    {
+        return new BulkInsertOperation.BulkInsertCommand(OperationId, _streamExposerPerShard[shardNumber], null, _skipOverwriteIfUnchanged);
+    }
+
+    protected override bool HasStream => _requestBodyStreamPerShard != null;
+
+    protected override Task WaitForId()
+    {
+        return Task.CompletedTask;
+    }
+
+    public override async Task StoreAsync(ShardedBatchCommandData command, string id)
```

Comment:
This is dangerous, no?
What if we have a 2 minutes run with only writes to `$0`, the `$1` and `$2` will timeout, no?

Reply:
Current implementation is not supposed to deal with that. As discussed with @karmeli87 and @ppekrol we might switch to HTTP/2 that will keep the connection alive out of the box.